examples/deepcopy-gen/generators/deepcopy.go

```diff
@@ -292,21 +292,40 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
 // for a type T is:
 //    func (t T) DeepCopy() T
 // or:
-//    func (t *T) DeepCopy() T
+//    func (t *T) DeepCopy() *T
 func hasDeepCopyMethod(t *types.Type) bool {
-	for mn, mt := range t.Methods {
-		if mn != "DeepCopy" {
-			continue
-		}
-		if len(mt.Signature.Parameters) != 0 {
-			return false
-		}
-		if len(mt.Signature.Results) != 1 || mt.Signature.Results[0].Name != t.Name {
-			return false
-		}
-		return true
+	f, found := t.Methods["DeepCopy"]
+	if !found {
+		return false
 	}
-	return false
+	if len(f.Signature.Parameters) != 0 {
+		return false
+	}
+	if len(f.Signature.Results) != 1 || f.Signature.Results[0].Name != t.Name {
+		return false
+	}
+	return true
+}
+
+// hasDeepCopyIntoMethod returns true if an appropriate DeepCopyInit() method is
+// defined for the given type.  This allows more efficient deep copy
+// implementations to be defined by the type's author.  The correct signature
+// for a type T is:
+//    func (t T) DeepCopyInto(t *T)
+// or:
+//    func (t *T) DeepCopyInit(t *T)
+func hasDeepCopyIntoMethod(t *types.Type) bool {
+	f, found := t.Methods["DeepCopyInto"]
+	if !found {
+		return false
+	}
+	if len(f.Signature.Parameters) != 1 {
+		return false
+	}
+	if len(f.Signature.Results) != 0 {
```

Comment:
shouldn't this also be checking the type of the argument (kind == pointer, etc)?

Reply:
Actually, I tend to remove all these checks. IMO it's better to create code that is not type-correct than skipping a DeepCopy functions that the user expects to be called, but it isn't because there is slight type-error. If we don't want to rely on the type checker, we could also error out here in the generator. But skipping slightly off funcs leads to hard to understand wrong behaviour of the generated code.

Reply:
I added another commit with fatal error. I would very much prefer that. If this turns out to be a problem on any external type with DeepCopy/Into, we can still add an opt-out tag. But we haven't seen such a case yet, so let's keep it simple for now.